File: python/Group_Anagrams.py
# https://leetcode.com/problems/group-anagrams/description/

from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = defaultdict(list)

        for word in strs:
            # 정렬한 값을 키로하여 딕셔너리에 추가
            anagrams[''.join(sorted(word))].append(word)
        return list(anagrams.values())
        
        '''
        sorted_strs = [''.join(sorted(list(word))) for word in strs]

        sorted_dict = dict()
        for idx, sorted_str in enumerate(sorted_strs):
            if sorted_str in sorted_dict:
                sorted_dict[sorted_str].append(idx)
            else:
                sorted_dict[sorted_str] = [idx]
        
        ret = []
        for index_list in sorted_dict.values():
            ret.append([strs[index] for index in index_list])
        return ret
        '''

File: java/leetcode/GroupAnagrams.java
package leetcode;

//https://leetcode.com/problems/group-anagrams/description/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<String, List<String>>();

        for (String str : strs) {
            char[] arr = str.toCharArray();
            Arrays.sort(arr);
            String key = new String(arr);

            List<String> list = map.getOrDefault(key, new ArrayList<String>());
            list.add(str);
            map.put(key, list);
        }

        return new ArrayList<List<String>>(map.values());
    }
}